### ISSUE TITLE
[5.2 backport candidate] Conversion of timestamp column to JSON and back fails

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#16575]
This issue was created just for the purpose of backport consideration - and had already been fixed.

This bug was already fixed by pull-request https://github.com/scylladb/scylladb/pull/14726 without realizing it. It was believed that that pull request only fixed two trivial bugs that weren't worthy of backporting - so it wasn't backported. But **this** bug, which the same PR fixed, is more serious, and we actually had a customer encounter it.

This bug is that if you convert a **timestamp** column to JSON (with SELECT JSON or toJson()) and then convert that back, the conversion back failed. We had a customer which did this: dump data to JSON and then attempting unsuccessfully to read it back.

The problem was that prior to that pull request, we wrote timestamps like `2023-08-10T10:10:46.497000` with **six digits** for the fractions of second, the last three being zero. This doesn't make sense (timestamps only have miilisecond precision), but the bigger problem is that Scylla refuses to read them back - it says:
```
marshaling error: unable to parse date '2014-01-01 12:15:45.0000000Z': marshaling error: Milliseconds length exceeds expected (7)
```
After that pull request, the output format becomes `2023-08-10 10:10:46.497` (note just 3 digits for milliseconds) which Scylla can read back.

Note that if the pull request is backported, some dtests which enshrined the wrong format may also need to be backported.

CC @alezzqz @ptrsmrn

**Comment by mykaul** [comment]
To which version we'd like / need to backport it? 5.4 and 5.2?

**Comment by nyh** [comment]
The fix (https://github.com/scylladb/scylladb/commit/ff721ec3e32610ac0a65113f20e20efd8c784eed) is already in 5.4, so we're left just with considering a backport to 5.2.

**Comment by nyh** [comment]
By the way, what scares me about this backport and why I'm not running to do it right away, is that it doesn't just accept the six-digit format we wrongly wrote in the past (this would have been completely safe) - it changes the output format to a three-digit format. In theory (though probably not in practice), someone might be relying on that wrong output format, and be surprised when it changes.

But perhaps we should do this backport anyway. After all, we already decided to do this in the Enterprise branches.

**Comment by ptrsmrn** [comment]
I think the other way round, i.e. having less precise timestamp, shouldn't cause harm, though you can never be sure. Even if, it's a minor change, and it should be enough to document it (will be in documented in changelog, I believe). So I would proceed with the backport.

**Comment by nyh** [comment]
Oh, another scary thing about this backport is that it will probably also require dtest backports, to avoid the wrong tests :-(

**Comment by mykaul** [comment]
Do we wish to do this backport? I suspect we just had a dtest failure in 5.2 that might be related (https://jenkins.scylladb.com/job/scylla-5.2/job/dtest-release/47/testReport/json_test/TestsJsonFullRowInsertSelect/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_simple_schema/ ) ?

**Comment by nyh** [comment]
I tried very much to avoid this backport (hoping that 5.2 would disappear ;-)) because I was afraid of the dtest problems, but I can do the backport now.

That being said, the dtest failures that this issue can cause or not cause are not random - they should happen each and every time. So how did this test start failing now? I suspect somebody backported some dtest changes without backporting the code changes. Maybe @fruch remembers something - we had a lot of issues about testing time formats in the last year.

Anyway, I'll backport the code changes and then take a look at the dtest to see what is going on.

**Comment by nyh** [comment]
Did the backport, commit 32be38dae556936b56061a3860c1da077602f9b4 in next-5.2.
I'm not sure if this issue will be closed by the Fixes appearing on a non-master branch. If not, this issue should be closed when that backport promotes to branch-5.2.
